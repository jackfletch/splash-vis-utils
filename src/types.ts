export interface Point {
  x: number;
  y: number;
}

export interface Shot {
  id: number;
  game_id: number;
  game_event_id: number | string;
  player_id: number;
  distance: number;
  x: number;
  y: number;
  made_flag: boolean;
}
