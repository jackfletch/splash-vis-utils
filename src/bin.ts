import {isLessThanDistance} from "./filters";
import {LeftRight, Shot} from "./types";

export interface Bin {
  made: number;
  total: number;
}

function bin(shots: Shot[], maxDistance: number): Bin[] {
  const bins: Bin[] = Array(maxDistance + 1)
    .fill(undefined)
    .map(() => ({made: 0, total: 0}));

  shots.filter(isLessThanDistance(maxDistance)).forEach(shot => {
    const {distance, made} = shot;
    bins[distance].total += 1;
    bins[distance].made += +made;
  });

  return bins;
}

export function binLeftRight(
  shots: Shot[],
  maxDistance: number
): LeftRight<Bin[]> {
  const binsLeft: Bin[] = Array(maxDistance + 1)
    .fill(undefined)
    .map(() => ({made: 0, total: 0}));
  const binsRight: Bin[] = Array(maxDistance + 1)
    .fill(undefined)
    .map(() => ({made: 0, total: 0}));

  shots.filter(isLessThanDistance(maxDistance)).forEach(shot => {
    const {distance, made, x} = shot;
    if (x < 0) {
      binsLeft[distance].total += 1;
      binsLeft[distance].made += +made;
    } else if (x > 0) {
      binsRight[distance].total += 1;
      binsRight[distance].made += +made;
    }
  });

  return {left: binsLeft, right: binsRight};
}

export default bin;
