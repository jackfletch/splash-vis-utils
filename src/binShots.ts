import {Point, Shot} from "./types";

interface ShotDistanceMap {
  [index: number]: Shot[];
}

interface BinnedShots {
  binData: Point[];
  totalShots: number;
  totalMakes: number;
}

function binShots(data: Shot[], maxDistance: number): BinnedShots {
  const binnedShots: ShotDistanceMap = {};
  let totalMakes = 0;
  let totalShots = data.length;

  data.forEach(shot => {
    if (shot.distance <= maxDistance) {
      if (shot.distance in binnedShots) {
        binnedShots[shot.distance].push(shot);
      } else {
        binnedShots[shot.distance] = [shot];
      }
      if (shot.made_flag) {
        totalMakes += 1;
      }
    } else {
      totalShots -= 1;
    }
  });

  const binData: Point[] = [];
  let x = 0;

  Object.keys(binnedShots)
    .map(Number)
    .forEach(bin => {
      while (x !== bin) {
        binData.push({x, y: 0});
        x += 1;
      }
      const shots = binnedShots[bin];
      binData.push({x: bin, y: shots.length});
      x += 1;
    });

  const binnedData = {
    binData,
    totalShots,
    totalMakes,
  };
  return binnedData;
}

export default binShots;
