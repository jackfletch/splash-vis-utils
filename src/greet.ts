export interface Person {
  firstName?: string;
  lastName: string;
}

function greet(person: Person): void {
  console.log("hello " + person?.firstName);
}

export default greet;
