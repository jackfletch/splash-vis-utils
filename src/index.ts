export {default as bin, binLeftRight} from "./bin";
export {default as binShots} from "./binShots";
export {default as ribbonShots} from "./ribbonShots";
export {Point, Shot} from "./types";
