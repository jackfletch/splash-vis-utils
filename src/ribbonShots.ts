import binByDistance from "./bin";
import {Shot} from "./types";

interface RibbonBin {
  shootingPct: number;
  width: number;
}

function ribbonShots(data: Shot[], maxDistance: number): RibbonBin[] {
  const bins = binByDistance(data, maxDistance);
  const totalShots = bins.reduce((a, b) => a + b.total, 0);

  const ribbonBins = bins.map((bin) => {
    const shootingPct = bin.total > 0 ? bin.made / bin.total : 0;
    const width = bin.total / totalShots;
    return {
      shootingPct,
      width,
    };
  });

  // TODO: implement smoothing function

  return ribbonBins;
}

export default ribbonShots;
