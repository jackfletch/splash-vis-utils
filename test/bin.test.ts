import {bin} from "../src";
import {DEFAULT_DISTANCE, MAX_DISTANCE, SMALL_DISTANCE} from "./constants";
import shot2544 from "./fixtures/shots/2544.json";

describe("bin", () => {
  it("Should correctly bin shots", () => {
    const result = bin(shot2544, DEFAULT_DISTANCE);
    expect(result).toMatchSnapshot("default");
  });

  it(`Should correctly bin shots with max distance of ${MAX_DISTANCE}`, () => {
    const result = bin(shot2544, MAX_DISTANCE);
    expect(result).toMatchSnapshot("max");
  });

  it(`Should correctly bin shots with max distance of ${SMALL_DISTANCE}`, () => {
    const result = bin(shot2544, SMALL_DISTANCE);
    expect(result).toMatchSnapshot("5");
  });

  it("Should correctly bin shots given only one shot", () => {
    const shots = [
      {
        id: 100442,
        game_id: 21600001,
        game_event_id: "8",
        player_id: 2544,
        distance: 11,
        x: -79,
        y: 80,
        made_flag: true,
      },
    ];
    const result = bin(shots, DEFAULT_DISTANCE);
    expect(result).toMatchSnapshot("default");
  });
});
