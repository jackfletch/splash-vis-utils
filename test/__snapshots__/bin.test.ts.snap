// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bin Should correctly bin shots given only one shot 1`] = `
Array [
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 1,
    "total": 1,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
]
`;

exports[`bin Should correctly bin shots with max distance of 5 1`] = `
Array [
  Object {
    "made": 270,
    "total": 316,
  },
  Object {
    "made": 132,
    "total": 180,
  },
  Object {
    "made": 52,
    "total": 87,
  },
  Object {
    "made": 14,
    "total": 32,
  },
  Object {
    "made": 9,
    "total": 18,
  },
  Object {
    "made": 8,
    "total": 25,
  },
]
`;

exports[`bin Should correctly bin shots with max distance of 100 1`] = `
Array [
  Object {
    "made": 270,
    "total": 316,
  },
  Object {
    "made": 132,
    "total": 180,
  },
  Object {
    "made": 52,
    "total": 87,
  },
  Object {
    "made": 14,
    "total": 32,
  },
  Object {
    "made": 9,
    "total": 18,
  },
  Object {
    "made": 8,
    "total": 25,
  },
  Object {
    "made": 6,
    "total": 18,
  },
  Object {
    "made": 10,
    "total": 19,
  },
  Object {
    "made": 4,
    "total": 14,
  },
  Object {
    "made": 3,
    "total": 15,
  },
  Object {
    "made": 3,
    "total": 14,
  },
  Object {
    "made": 10,
    "total": 19,
  },
  Object {
    "made": 10,
    "total": 20,
  },
  Object {
    "made": 11,
    "total": 30,
  },
  Object {
    "made": 12,
    "total": 26,
  },
  Object {
    "made": 16,
    "total": 32,
  },
  Object {
    "made": 8,
    "total": 28,
  },
  Object {
    "made": 4,
    "total": 18,
  },
  Object {
    "made": 5,
    "total": 14,
  },
  Object {
    "made": 11,
    "total": 23,
  },
  Object {
    "made": 7,
    "total": 20,
  },
  Object {
    "made": 5,
    "total": 23,
  },
  Object {
    "made": 7,
    "total": 20,
  },
  Object {
    "made": 8,
    "total": 23,
  },
  Object {
    "made": 27,
    "total": 74,
  },
  Object {
    "made": 36,
    "total": 119,
  },
  Object {
    "made": 28,
    "total": 62,
  },
  Object {
    "made": 9,
    "total": 19,
  },
  Object {
    "made": 5,
    "total": 14,
  },
  Object {
    "made": 2,
    "total": 9,
  },
  Object {
    "made": 2,
    "total": 4,
  },
  Object {
    "made": 0,
    "total": 2,
  },
  Object {
    "made": 1,
    "total": 3,
  },
  Object {
    "made": 1,
    "total": 2,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 1,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
]
`;

exports[`bin Should correctly bin shots with max distance of 35 1`] = `
Array [
  Object {
    "made": 270,
    "total": 316,
  },
  Object {
    "made": 132,
    "total": 180,
  },
  Object {
    "made": 52,
    "total": 87,
  },
  Object {
    "made": 14,
    "total": 32,
  },
  Object {
    "made": 9,
    "total": 18,
  },
  Object {
    "made": 8,
    "total": 25,
  },
  Object {
    "made": 6,
    "total": 18,
  },
  Object {
    "made": 10,
    "total": 19,
  },
  Object {
    "made": 4,
    "total": 14,
  },
  Object {
    "made": 3,
    "total": 15,
  },
  Object {
    "made": 3,
    "total": 14,
  },
  Object {
    "made": 10,
    "total": 19,
  },
  Object {
    "made": 10,
    "total": 20,
  },
  Object {
    "made": 11,
    "total": 30,
  },
  Object {
    "made": 12,
    "total": 26,
  },
  Object {
    "made": 16,
    "total": 32,
  },
  Object {
    "made": 8,
    "total": 28,
  },
  Object {
    "made": 4,
    "total": 18,
  },
  Object {
    "made": 5,
    "total": 14,
  },
  Object {
    "made": 11,
    "total": 23,
  },
  Object {
    "made": 7,
    "total": 20,
  },
  Object {
    "made": 5,
    "total": 23,
  },
  Object {
    "made": 7,
    "total": 20,
  },
  Object {
    "made": 8,
    "total": 23,
  },
  Object {
    "made": 27,
    "total": 74,
  },
  Object {
    "made": 36,
    "total": 119,
  },
  Object {
    "made": 28,
    "total": 62,
  },
  Object {
    "made": 9,
    "total": 19,
  },
  Object {
    "made": 5,
    "total": 14,
  },
  Object {
    "made": 2,
    "total": 9,
  },
  Object {
    "made": 2,
    "total": 4,
  },
  Object {
    "made": 0,
    "total": 2,
  },
  Object {
    "made": 1,
    "total": 3,
  },
  Object {
    "made": 1,
    "total": 2,
  },
  Object {
    "made": 0,
    "total": 0,
  },
  Object {
    "made": 0,
    "total": 0,
  },
]
`;
